File: ga/gpt3_5_result_2024_6/result/message_queue_gpt/9_message_queue.c
use std::mem;
use std::ffi::CString;
use std::os::raw::c_char;
use std::os::raw::c_int;
use libc::{msgget, msgsnd, msgrcv, IPC_PRIVATE, IPC_EXCL};
use std::ptr;
use std::process::Command;

const MY_TYPE: i64 = 9527;

struct MsgBuf {
    mtype: i64,
    mtext: [u8; 100],
    number: i32,
}

fn main() {
    let msgid: c_int;
    let mut pid: i32;

    unsafe {
        msgid = msgget(IPC_PRIVATE, 0o666 | IPC_EXCL);
        if msgid == -1 {
            println!("Error creating message queue");
            return;
        }

        pid = Command::new("sh")
            .arg("-c")
            .arg("echo $$")
            .output()
            .expect("Failed to execute command")
            .stdout
            .parse()
            .expect("Failed to parse output");

        if pid > 0 {
            std::thread::sleep(std::time::Duration::from_secs(1));

            let mut buff = MsgBuf {
                mtype: MY_TYPE,
                mtext: [0; 100],
                number: 0,
            };

            println!("Please enter a string you want to send:");
            let mut input_text = String::new();
            std::io::stdin().read_line(&mut input_text).expect("Failed to read line");
            let c_input_text = CString::new(input_text.trim()).expect("Failed to convert string to CString");
            let c_input_text_ptr = c_input_text.as_ptr() as *mut c_char;
            ptr::copy_nonoverlapping(c_input_text_ptr, buff.mtext.as_mut_ptr(), input_text.len());

            println!("Please enter a number you want to send:");
            let mut input_number = String::new();
            std::io::stdin().read_line(&mut input_number).expect("Failed to read line");
            buff.number = input_number.trim().parse().expect("Failed to parse number");

            msgsnd(msgid, &buff as *const _ as *const std::ffi::c_void, mem::size_of::<MsgBuf>() - mem::size_of::<i64>(), 0);

            waitpid(pid, std::ptr::null_mut(), 0);
        } else if pid == 0 {
            let mut buff = MsgBuf {
                mtype: MY_TYPE,
                mtext: [0; 100],
                number: 0,
            };
            
            println!("Child process is waiting for message:");
            msgrcv(msgid, &mut buff as *mut _ as *mut std::ffi::c_void, mem::size_of::<MsgBuf>() - mem::size_of::<i64>(), MY_TYPE, 0);
            let received_text = std::str::from_utf8(&buff.mtext).expect("Failed to convert text to UTF-8");
            println!("Child process read from message: {}, {}", received_text, buff.number);
        } else {
            println!("Error forking process");
        }
    }
}

File: ga/glm4_result_2024_6/result/message_queue_glm/9_message_queue.rs
use std::ffi::CString;
use std::io::{self, Read};
use std::os::raw::c_long;
use std::ptr;
use std::thread;
use std::time::Duration;
use std::wait::WaitStatus;

use libc::{fork, gets, msgget, msgrcv, msgsnd, IPC_PRIVATE, IPC_EXCL, IPC_CREAT, IPC_NOWAIT, waitpid};

const MY_TYPE: c_long = 9527;

fn main() {
    let msgid = unsafe { msgget(IPC_PRIVATE, IPC_CREAT | IPC_EXCL | 0o666) };

    if msgid == -1 {
        eprintln!("msgget failed");
        return;
    }

    match unsafe { fork() } {
        0 => {
            // Child process
            let mut buff = libc::msgbuf {
                mtype: MY_TYPE,
                mtext: [0; 100],
                number: 0,
            };

            let _ = thread::sleep(Duration::from_secs(1));

            let _ = io::stdin().read_line(&mut buff.mtext);
            let _ = io::stdin().read_line(&mut buff.number);

            if unsafe { msgsnd(msgid, &buff, (std::mem::size_of::<libc::msgbuf>() - std::mem::size_of::<c_long>()) as i32, IPC_NOWAIT) } == -1 {
                eprintln!("msgsnd failed");
            }
        }
        pid if pid > 0 => {
            // Parent process
            let mut buff = libc::msgbuf {
                mtype: MY_TYPE,
                mtext: [0; 100],
                number: 0,
            };

            if unsafe { msgrcv(msgid, &mut buff, (std::mem::size_of::<libc::msgbuf>() - std::mem::size_of::<c_long>()) as i32, MY_TYPE, 0) } == -1 {
                eprintln!("msgrcv failed");
            } else {
                println!("Parent process read from msg: {}", buff.mtext);
            }

            let _ = unsafe { waitpid(pid, ptr::null_mut(), 0) };
        }
        _ => {
            eprintln!("fork failed");
        }
    }
}